### PR TITLE
Update ILLink wording

### PR DESCRIPTION
**Pull request by tlakollo**
Runtime repo uses the term linker to refer to the native linker compiler, in an attempt for disambiguation for the term linker, and given that now the package is being produced in runtime under src/tools/illink. This is a PR to rename some of the linker instances and replace them with the ILLink, illink, IL Linker, trimming, or trimmer terms.

In this PR:
* Update linker wording to disambiguate the term from native linker, switched to ILLink, illink, IL Linker, trimming, or trimmer
* Deleted illink steps-explained document since is no longer accurate
* This change did not change attribute justification messages

### CODE REVIEW

File: src/coreclr/tools/aot/ILLink.Shared/DiagnosticId.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ public enum DiagnosticId
 		CouldNotFindAssemblyReference = 1009,
 		CouldNotLoadAssembly = 1010,
 		FailedToWriteOutput = 1011,
-		LinkerUnexpectedError = 1012,
+		ILLinkUnexpectedError = 1012,
```
Comment (agocke) [inline src/coreclr/tools/aot/ILLink.Shared/DiagnosticId.cs:25]: This is a public API

**Review by agocke — CHANGES_REQUESTED** [review]

File: src/coreclr/tools/aot/ILLink.Shared/DiagnosticId.cs (hunk 1 above)
Comment (tlakollo) [reply src/coreclr/tools/aot/ILLink.Shared/DiagnosticId.cs:25]: I feel like part of maintaining the DiagnosticId's is something that inevitably we have to do, there have been diagnosticId clashes in the past or ids that we no longer use and rename as `_unused_`. In fact, I was planning to, later on, fix some of the grammar problems in some of the DiagnosticId's (MissingArgumentForCommanLineOptionName = 1018, CannotConverValueToType = 1042, XmlUnkownBodyModification = 2011, XmlCouldNotFindGetAccesorOfPropertyOnType = 2018, XmlCouldNotFindSetAccesorOfPropertyOnType = 2019, XmlMoreThanOneValyForParameterOfMethod = 2024, XmlElementDoesNotContainRequiredAttributeFullname = 2029, XmlUnsuportedWildcard = 2100) which were flagged by the Visual Studio spellcheck feature.
I see this change as an improvement to readability which kind of falls in the same bucket for me, not so long ago we modify the SharedStrings to reflect reporting an illink issue in runtime when this Id is used.

Kind of unrelated I also noticed that this error id is used by dataflow, a piece that is reused by NativeAOT and although the message is not about opening an issue in runtime with the tag ILLink, I think we shouldn't be using this ID in other parts besides the Driver file in illink.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ClonedConstructedEETypeNode.cs [hunk 2]
```diff
@@ -17,7 +17,7 @@ public ClonedConstructedEETypeNode(NodeFactory factory, TypeDesc type) : base(fa
         public override ISymbolNode NodeForLinkage(NodeFactory factory) => this;
 
         //
-        // A cloned type must be named differently than the type it is a clone of so the linker
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ClonedConstructedEETypeNode.cs:20]: This refers to the real linker.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs [hunk 3]
```diff
@@ -721,7 +721,7 @@ public void EmitCFICodes(int offset)
                 if (_targetPlatform.IsOSXLike)
                 {
                     // Emit a symbol for beginning of the frame. This is workaround for ld64
-                    // linker bug which would produce DWARF with incorrect pcStart offsets for
+                    // ILLink tool bug which would produce DWARF with incorrect pcStart offsets for
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:724]: ld64 is a real linker.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs [hunk 4]
```diff
@@ -914,7 +914,7 @@ private bool ShouldShareSymbol(ObjectNode node)
             if (!(node is ISymbolNode))
                 return false;
 
-            // These intentionally clash with one another, but are merged with linker directives so should not be Comdat folded
+            // These intentionally clash with one another, but are merged with ILLink tool directives so should not be Comdat folded
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:917]: Real linker as well.

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEHeader.cs [hunk 5]
```diff
@@ -16,12 +16,12 @@ public sealed class PEHeader
         public PEMagic Magic { get; }
 
         /// <summary>
-        /// The linker major version number.
```
Comment (MichalStrehovsky) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEHeader.cs:19]: Real linker.

File: src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEHeader.cs [hunk 6]
```diff
@@ -16,12 +16,12 @@ public sealed class PEHeader
...
+        /// ILLink major version number.
         /// </summary>
         public byte MajorLinkerVersion { get; }
 
         /// <summary>
-        /// The linker minor version number.
```
Comment (MichalStrehovsky) [inline src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEHeader.cs:24]: Real linker.

File: src/libraries/System.Runtime/tests/TrimmingTests/System.Runtime.TrimmingTests.proj [hunk 7]
```diff
@@ -19,12 +19,6 @@
       <EnabledFeatureSwitches>System.Globalization.Invariant;System.Globalization.PredefinedCulturesOnly</EnabledFeatureSwitches>
     </TestConsoleAppSourceFiles>
     <TestConsoleAppSourceFiles Include="StackFrameHelperTest.cs">
-      <!-- There is a bug with the linker where it is corrupting the pdbs while trimming
```
Comment (MichalStrehovsky) [inline src/libraries/System.Runtime/tests/TrimmingTests/System.Runtime.TrimmingTests.proj:22]: This is a good change, but orthogonal to renaming (it is a functional change). In general it's preferable not to do functional changes in big PRs like this.

File: src/tests/nativeaot/SmokeTests/UnitTests/Generics.cs [hunk 8]
```diff
@@ -61,7 +61,7 @@ internal static int Run()
 
     /// <summary>
     /// Tests that we properly track dictionary dependencies of generic methods.
-    /// (Getting this wrong is a linker failure.)
+    /// (Getting this wrong is a trimming failure.)
```
Comment (MichalStrehovsky) [inline src/tests/nativeaot/SmokeTests/UnitTests/Generics.cs:64]: Referring to real linker here.

File: src/tests/readytorun/tests/generics.cs [hunk 9]
```diff
@@ -297,7 +297,7 @@ static IEnumerable<T> CallGenVirtMethod<T>(NormalClass n, object o)
 
     /// <summary>
     /// Tests that we properly track dictionary dependencies of generic methods.
-    /// (Getting this wrong is a linker failure.)
+    /// (Getting this wrong is a trimming failure.)
```
Comment (MichalStrehovsky) [inline src/tests/readytorun/tests/generics.cs:300]: also real linker.

File: src/coreclr/tools/aot/ILLink.Shared/DiagnosticId.cs (hunk 1 above)
Comment (tlakollo) [reply src/coreclr/tools/aot/ILLink.Shared/DiagnosticId.cs]: I will see if I can rename it once I deal with the fact that Dataflow shouldn't produce this error id. For now I reverted the changes

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ClonedConstructedEETypeNode.cs [hunk 10]
```diff
@@ -17,7 +17,7 @@ public ClonedConstructedEETypeNode(NodeFactory factory, TypeDesc type) : base(fa
...
+        // A cloned type must be named differently than the type it is a clone or the linker
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ClonedConstructedEETypeNode.cs:20]: Not a native english speaker, but the previous sentence made sense. The update doesn't.

We can replace "the type it is a clone of" with "the original type" if it's more readable like that, but the "or" is not a good change.

**Review by MichalStrehovsky — APPROVED** [review]
I did not look at files under src/tools.

Also, there might be other references in user-visible places such as https://github.com/dotnet/runtime/blob/361ba6146e523a3f8b4a90cbadcefefae6bc7717/src/libraries/System.Text.Json/src/Resources/Strings.resx#L525

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ClonedConstructedEETypeNode.cs (hunk 10 above)
Comment (agocke) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ClonedConstructedEETypeNode.cs:20]: Yup, as a native speaker Michal's right -- the old wording or his proposed wording are both fine, but the change here is ungrammatical.

**Review by agocke — APPROVED** [review]
LGTM aside from Michal's comment

**Pull request merged**